src/codegen/tests/debug_tests.rs

```diff
@@ -342,3 +342,142 @@ fn switch_case_debug_info() {
     !9 = !DILocation(line: 0, scope: !4)
     "###);
 }
+
+#[test]
+fn dbg_declare_has_valid_metadata_references_for_methods() {
+    let codegen = codegen_with_debug_version(
+        r"
+        FUNCTION_BLOCK fb
+        METHOD foo
+        END_METHOD
+        END_FUNCTION_BLOCK
+        ",
+        4,
+    );
+
+    // We want to make sure the `dbg.declare` for the method `foo` references a non-empty metadata field, i.e.
+    // `!<number>`
+    assert!(codegen.contains(r#"call void @llvm.dbg.declare(metadata %fb.foo* %1, metadata !17, metadata !DIExpression()), !dbg !16"#));
+    assert!(codegen
+        .contains(r#"!17 = !DILocalVariable(name: "fb.foo", scope: !15, file: !2, line: 3, type: !18)"#));
+
+    // ...for reference, it would previously point to `!4` which is empty (`!4 = {}`). `!4` shouldn't exist in
```

Comment:
I think `!4` is not wrong here - it just points out that the struct `fb` does not have any members and therefore no local nodes need to be retained.
For comparison, this is an [example](https://llvm.org/docs/SourceLevelDebugging.html#object-lifetimes-and-scoping) by LLVM. Here, the function `foo` does not have any members, `!DISubprogram`'s `retainedNodes` metadata points to `!2` which is `!2 = !{}`.

Reply:
Yes, you're right. I was under the impression that the `!4` wasn't being used in the IR at all but that's not true.